server/kv.go

```diff
@@ -0,0 +1,199 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"fmt"
+	"math"
+	"path"
+	"time"
+
+	"github.com/coreos/etcd/clientv3"
+	"github.com/gogo/protobuf/proto"
+	"github.com/juju/errors"
+	"github.com/ngaut/log"
+	"github.com/pingcap/kvproto/pkg/metapb"
+	"golang.org/x/net/context"
+)
+
+const (
+	kvRangeLimit      = 100000
+	kvRequestTimeout  = time.Second * 10
+	kvSlowRequestTime = time.Second * 1
+)
+
+var (
+	errTxnFailed = errors.New("failed to commit transaction")
+)
+
+// kv wraps all kv operations, keep it stateless.
+type kv struct {
+	s           *Server
+	client      *clientv3.Client
+	clusterPath string
+}
+
+func newKV(s *Server) *kv {
+	return &kv{
+		s:           s,
+		client:      s.client,
+		clusterPath: path.Join(s.rootPath, "raft"),
+	}
+}
+
+func (kv *kv) txn(cs ...clientv3.Cmp) clientv3.Txn { return kv.s.leaderTxn(cs...) }
+
+func (kv *kv) storePath(storeID uint64) string {
+	return path.Join(kv.clusterPath, "s", fmt.Sprintf("%020d", storeID))
+}
+
+func (kv *kv) regionPath(regionID uint64) string {
+	return path.Join(kv.clusterPath, "r", fmt.Sprintf("%020d", regionID))
+}
+
+func (kv *kv) loadMeta(meta *metapb.Cluster) (bool, error) {
+	return kv.loadProto(kv.clusterPath, meta)
+}
+
+func (kv *kv) saveMeta(meta *metapb.Cluster) error {
+	return kv.saveProto(kv.clusterPath, meta)
+}
+
+func (kv *kv) loadStore(storeID uint64, store *metapb.Store) (bool, error) {
+	return kv.loadProto(kv.storePath(storeID), store)
+}
+
+func (kv *kv) saveStore(store *metapb.Store) error {
+	return kv.saveProto(kv.storePath(store.GetId()), store)
+}
+
+func (kv *kv) loadRegion(regionID uint64, region *metapb.Region) (bool, error) {
+	return kv.loadProto(kv.regionPath(regionID), region)
+}
+
+func (kv *kv) saveRegion(region *metapb.Region) error {
+	return kv.saveProto(kv.regionPath(region.GetId()), region)
+}
+
+func (kv *kv) loadStores(stores *storesInfo, rangeLimit int64) error {
+	nextID := uint64(0)
+	endStore := kv.storePath(math.MaxUint64)
+	withRange := clientv3.WithRange(endStore)
+	withLimit := clientv3.WithLimit(rangeLimit)
+
+	for {
+		key := kv.storePath(nextID)
+		resp, err := kvGet(kv.client, key, withRange, withLimit)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		if len(resp.Kvs) == 0 {
```

Comment:
if len(resp.Kvs) < rangeLimit, we can still return, no need to call another get.